ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+
+Workspace CRD serves as an abstraction to be able to make changes to a workspace or create new workspaces for a specific users. 
+
+This abstraction also enable Konflux to implement other features in the future like for example limmiting the amount of workspaces per user, allowing workspace delete self-service and adding workspace delete protection.
+
+
+### Upstream
+
+There is not support for multi-cluster for a single instance of Konflux. The Idea here is to require a new deployment of konflux of there is a need to scale beyond a single cluster. Therefore, there is a one to one relationship between konflux instance and k8s cluster.
+
+With this contraint in place workspace concept can be implemented differently
+We relay on the ability to create namespaces directly. Which requires access to all authenticated users to have premissions to create a namespace.
+
+The boostrapping of the namespace created is delegate to a GitOps workflow that will automatically create an application with the necesary resources, including RoleBinding and Quotas and will continue to synch this through the lifecycle of the project.
+
+To be able to see exisiting workspaces for a specific user, we developed a small abstraction called workspace manager, which is able to determine what namepsace have a rolebinding providing access to the current user.
+
+There is currently no way to delete a workspace self-service. User will have to make a request to a konflux admin to delete a specific workspace
+
+### Namespace Limitations
+
+Kubernetes namespaces where not designed for self-service provisioning in a multi-tenant.
+
+Access for namespace list and create operation is all namepaces or none. This affects multi-tenancy if self-service is desired.
+
+Access to get and delete can be granted via RBAC once the namespace is created. This will required admin access which also affect self-service.
+
+Allowing access to specific Roles or ClusterRoles via RoleBinding can be done by an Admin, also braking self-service. This is required to be able to execute Konflux actions inside the namespace.
+
+## Decision
+
+Introduce a workspace CRD and corresponding controller as a project of our community.
+
+We will provide all authenticated users access to create their own workspaces within a limit.
+They will also be able to update, list, get and delete only workspaces they own. 
+
+## Consequences
+
+Extending k8s API with the workspace concept will allow konflux user to manage their konflux workspaces in a secured, self-service manner while konflux admin can maintain multi-tenancy.
+
+Such extention of the API that wraps namepsace has precedence, for example OpenShift apis provide a Project kind the wraps a k8s namespace.
+
+### Additional benefit 
+
+We can automate the bootstraping of namespaces with extra resources a konflux admin deems proper for their specific use case.
```

Comment:
This is a good feature that can also achieved by existing projects such as kyverno that knows how to do it for namespaces.

Reply:
I agree, there are multiple solutions to a problem, this is not an exception

Reply:
Is this something you really want to provide as part of the upstream project? Let's not talk about tools (KubeSaw, Kyverno, ArgoCD, ...), but about requirements for the feature set of the upstream project. 

It's definitely needed for the internal RH instance of Konflux, but this is completely different topic which should be discussed separately.